### PR TITLE
File format security issues

### PR DESCRIPTION
**Pull request by derobins**
Addresses file format security issues detected via fuzzing.

Credit: Amazon Web Services and @qkoziol

### CODE REVIEW

**Comment by derobins** [comment]
This needs RELEASE.txt info, which we'll add shortly. Let's get this review started so we can get this into develop and hdf5_1_14 ASAP.

File: src/H5Spoint.c
Comment (mattjala) [inline src/H5Spoint.c:586]: How is this allocation not causing constant memory errors in the current version of the library due to being too small?

Comment (jhendersonHDF) [reply src/H5Spoint.c:586]: It's likely due to memory that was already allocated on the free list being reused here

**Comment by derobins** [comment]
I have a set of files to add to https://github.com/HDFGroup/cve_hdf5, which I will do after we merge these changes to hdf5_1_14. If I do it now, the CI will break.

File: src/H5Odtype.c
Comment (jhendersonHDF) [inline src/H5Odtype.c:492]: Seems like this should be `dt->shared->u.compnd.nmembs - 1`

File: src/H5Olinfo.c
Comment (jhendersonHDF) [inline src/H5Olinfo.c]: Can be printed with `message: %" PRId64` here rather than casting

**Review by mattjala — APPROVED** [review]

File: test/tmisc.c
Comment (jhendersonHDF) [inline test/tmisc.c]: Should probably use "tmisc38c.h5" for this filename just to avoid any test weirdness in the future

Comment (byrnHDF) [reply test/tmisc.c]: Oh, it will overwrite the B file and not create a C, right?

File: tools/test/h5dump/expected/tnbit.ddl
Comment (jhendersonHDF) [inline tools/test/h5dump/expected/tnbit.ddl:7]: Likely still want to leave the size and compression ratio masked out here

File: test/tmisc.c
Comment (jhendersonHDF) [reply test/tmisc.c]: Looks like it from the test

File: src/H5Pfapl.c
Comment (jhendersonHDF) [inline src/H5Pfapl.c:6340]: Not a necessity, but it'd be nice to account for `H5P_DEFAULT` here

File: src/H5Ppublic.h
Comment (jhendersonHDF) [inline src/H5Ppublic.h]: 1.14.4

File: src/H5Ppublic.h
Comment (jhendersonHDF) [inline src/H5Ppublic.h]: Also 1.14.4

File: test/tmisc.c
Comment (jhendersonHDF) [inline test/tmisc.c]: Should move this outside the if - else block so we close `tid` even if `H5Tcommit2` fails above

**Review by qkoziol — APPROVED** [review]

File: src/H5Tcommit.c
Comment (bmribler) [inline src/H5Tcommit.c:547]: Can oloc_ptr->addr still be used after H5O_close(oloc_ptr, NULL)?

File: src/H5Tconv.c
Comment (jhendersonHDF) [inline src/H5Tconv.c:3839]: Rather than hiding the const warnings here due to `H5T_vlen_reclaim_elmt` taking a `H5T_t *` parameter, I think we can just update `H5T_vlen_reclaim_elmt` to take a `const H5T_t *` instead. All it does is pass that parameter to `H5T__vlen_reclaim`, which takes a `const H5T_t *`. Would be much cleaner.

File: src/H5Tconv.c
Comment (jhendersonHDF) [inline src/H5Tconv.c]: Same comment here about `H5T_vlen_reclaim_elmt`

File: src/H5Tconv.c
Comment (jhendersonHDF) [inline src/H5Tconv.c:4299]: After looking at this and the above side by side, it looks like the only difference is whether we call `H5T_vlen_reclaim_elmt` or `H5Rdestroy` on the element. Would be much nicer inside a helper function.

File: src/H5Tconv.c
Comment (jhendersonHDF) [inline src/H5Tconv.c]: Nitpick: this is for freeing variable-length elements rather than destroying references. Same for below

File: src/H5private.h
Comment (jhendersonHDF) [inline src/H5private.h]: This is the only part of the PR I keep getting hung up on. If a large enough value is passed in for `size`, won't this be implementation-defined what happens if it overflows when casting to `intptr_t`? It seems like we need something like:

`((size <= (size_t)INTPTR_MAX) && ((intptr_t)(size) < 0))`

though, I'm not quite sure that's right either in the case where `INTPTR_MAX` is larger than `SIZE_MAX`, causing wraparound when casting `INTPTR_MAX` to `size_t`

File: src/H5T.c
Comment (jhendersonHDF) [inline src/H5T.c:6656]: Testing on PPC64LE by @hyoklee shows that this probably won't work correctly for the IBM long double type. Suggest either:

```C
#if LDBL_MANT_DIG == 106
    /* This won't work for the IBM long double type */
    HGOTO_DONE(false);
#else
    ...
#endif
```

or some other solution, until we can maybe address issues with the datatype.

File: src/H5Ppublic.h
Comment (derobins) [reply src/H5Ppublic.h]: Fixed

File: src/H5Ppublic.h
Comment (derobins) [reply src/H5Ppublic.h]: Fixed

File: tools/test/h5dump/expected/tnbit.ddl
Comment (derobins) [reply tools/test/h5dump/expected/tnbit.ddl:7]: Fixed

File: src/H5Olinfo.c
Comment (derobins) [reply src/H5Olinfo.c]: Fixed

File: src/H5Tconv.c
Comment (derobins) [reply src/H5Tconv.c:3839]: Fixed

File: src/H5Tconv.c
Comment (derobins) [reply src/H5Tconv.c]: Fixed

File: src/H5Tconv.c
Comment (derobins) [reply src/H5Tconv.c]: Fixed

File: src/H5Tconv.c
Comment (derobins) [reply src/H5Tconv.c:4299]: Agree, but I'll leave this for later work. Right now I just want to get this in.

File: src/H5Tcommit.c
Comment (derobins) [reply src/H5Tcommit.c:547]: No, it points to temp_oloc, which is local and cleaned up immediately after this.

File: test/tmisc.c
Comment (derobins) [reply test/tmisc.c]: Done

File: test/tmisc.c
Comment (derobins) [reply test/tmisc.c]: Fixed

File: src/H5private.h
Comment (fortnern) [reply src/H5private.h]: So I think it's technically possible to have a valid buffer so large that ptrdiff_t wraps to negative when computing the size. This could be an issue on 32 bit systems that are pushing the limits of their architecture. So size could be negative when cast to ptrdiff_t and still be valid. I don't know if the rest of the library would handle this case correctly.

Comment (fortnern) [reply src/H5private.h]: Maybe something like `(((ptr) > (buffer_end)) || (((ptr) + (size) - 1) > (buffer_end)) || (((ptr) + (size)) < (ptr)))`

We might want to cast the pointers to `const uint8_t *` as well, and size to `size_t`.

File: src/H5Odtype.c
Comment (jhendersonHDF) [reply src/H5Odtype.c:492]: Actually this one's fine. `dt->shared->u.compnd.nmembs` is the loop index here so this should be valid.

File: src/H5private.h
Comment (jhendersonHDF) [reply src/H5private.h]: Hmm yes I think this is a bit more in line with what I was thinking.

Comment (jhendersonHDF) [reply src/H5private.h]: After a little research though, it seems some compilers often remove a `ptr + len < ptr` type check since it's UB anyway.

Comment (qkoziol) [reply src/H5private.h]: I suggest getting my notes for files that tripped overflows from Dana and making certain that any change to what I worked out last year are still protecting from the related corrupted files.  Some of them were pretty obscure... :-/

Comment (jhendersonHDF) [reply src/H5private.h]: Definitely plan to test against the CVE files. I think it's mostly a matter of needing to check `size` against the max of the signed type it's being cast into first, and then casting it and checking against < 0 before then doing the pointer subtraction and comparison. Unfortunately it's difficult to do this in a nice way since the macro has no type information and it's not always a `size_t` value that's being passed in.

File: src/H5encode.h
Comment (jhendersonHDF) [inline src/H5encode.h:227]: Casting this back to a `uint64_t` seems to be showing a warning:

```
conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Warith-conversion]
  227 |             n = (uint64_t)(((uint64_t)n << 8) | *(--p));                                                     \
      |                 ^
src/H5Olayout.c:398:33: note: in expansion of macro ‘H5F_DECODE_LENGTH’
  398 |                                 H5F_DECODE_LENGTH(f, p, mesg->storage.u.chunk.u.single.nbytes);
      |                                 ^~~~~~~~~~~~~~~~~
```

Seems like we could end up trying to UINT64DECODE into a `uint32_t` value

File: src/H5private.h
Comment (jhendersonHDF) [reply src/H5private.h]: Rewrote this slightly. It's basically the same thing, but using `ptrdiff_t` instead of `intptr_t`, checking against `PTRDIFF_MAX` before casting and also doing the check for negative values before the typical overflow check. Passes all current CVE tests, including the ones specifically around the changes in this PR.

Because of the "type-limits" warnings that had to be suppressed in this PR, I'm tempted to just move the macro into an inline function, but for the sake of getting these changes in, I don't think needs to happen right now.

File: src/H5encode.h
Comment (jhendersonHDF) [reply src/H5encode.h:227]: Updated this to decode into a uint64_t and then checked_assign it to the uint32_t field.

**Comment by jhendersonHDF** [comment]
CMake seems to be failing in the `H5DUMP-tnbit` test because the compression ratio line isn't getting masked out for some reason. Not sure on netcdf failures, but investigating.

**Comment by jhendersonHDF** [comment]
Updating the `H5_IS_KNOWN_BUFFER_OVERFLOW` macro to use `H5_IS_BUFFER_OVERFLOW` when `skip` is false caused the netCDF tests to start failing in one test case when trying to open a file after creating a dataset with an opaque type in it. This fails both before and after my changes to `H5_IS_BUFFER_OVERFLOW`. More investigation is likely needed there, but I'll put the change back for now.

File: src/H5Odtype.c
Comment (fortnern) [inline src/H5Odtype.c:183]: Could we support this as a no-op?

File: src/H5Odtype.c
Comment (fortnern) [inline src/H5Odtype.c:185]: Could we get rid of the -1 and change the >= to a > here (and in other places in this function)?

File: src/H5Odtype.c
Comment (fortnern) [inline src/H5Odtype.c:492]: I think we should either always use `offset + size > dt->shared->size` or `offset + size - 1 >= dt->shared->size`

File: src/H5Oattribute.c
Comment (fortnern) [inline src/H5Oattribute.c:215]: Maybe move the checksum part up so we can short circuit the more expensive function call

File: src/H5Odtype.c
Comment (fortnern) [inline src/H5Odtype.c:818]: This doesn't seem to suppress the error when checksumming is turned on

**Comment by derobins** [comment]
> CMake seems to be failing in the `H5DUMP-tnbit` test because the compression ratio line isn't getting masked out for some reason. Not sure on netcdf failures, but investigating.

In the Autotools, I had to change the lead digit of a macro, since it goes from 10 to 1.

**Comment by jhendersonHDF** [comment]
Will address the H5_IS_BUFFER_OVERFLOW issue and @fortnern's comments in a separate PR so as to not hold up this PR and create a bunch of extra testing work.

**Pull request merged**
